[RFC] clipboard: don't clobber "0 when deleting to unnamed

bfredl opened the pull request:
This tries to fix an issue from #1182 reported by #1806. I got the semantics wrong when doing an unnamed delete, because I somehow believed the "unnamed register" actually was a register. 
@pgdouyon does this fix the issue for you?
Probably more register tests should be added, so leaving this a WIP.

bfredl committed:
clipboard: don't clobber "0 when deleting to unnamed

pgdouyon commented:
Sorry for the delay, yes this branch fixes the clobbering for me.  Thanks for the quick fix, hopefully we can get this merged soon.  I'll take a look this weekend and see if I can help out with the tests.

bfredl committed:
clipboard: more register tests for delete

bfredl commented:
Think I have tested most of the code paths I've touched now, so marking RFC. 

> I'll take a look this weekend and see if I can help out with the tests.
More tests from someone with more experience with vim registers is certainly welcome :+1:

justinmk commented on test/functional/clipboard/clipboard_provider_spec.lua:
```diff
@@ -38,6 +38,26 @@ local function basic_register_test()
+  feed('gg"adwj"1P^"-P')
+  expect([[
+    , stuff and some more
+    some textsome some text, stuff and some more]])
```
I don't see `set clipboard=unnamed` in these tests, should we add duplicate tests for that case?

bfredl replied:
This code is in a function that's called twice, without/with unnamed

justinmk commented:
The CI failure is due to https://github.com/neovim/neovim/issues/1831

bfredl commented:
Repushed, now they pass.

bfredl commented:
Cleaned up some duplicate logic and added a `#define` for a special value (use latest register). Also, now it should be clear in the code that `clipboard=unnamed` doesn't affect the register semantics of delete.

justinmk commented:
behavior matches vim exactly, as far as what's in the functional tests. Nice work! merged.
